### PR TITLE
Detect simple redirect loops.

### PR DESCRIPTION
**Pull request by isoos**

### CODE REVIEW

File: app/lib/shared/handler_helpers.dart [hunk 1]
```diff
@@ -76,6 +77,50 @@ shelf.Handler wrapHandler(
...
+    final rs = await handler(request);
+    if (rs.statusCode >= 300 && rs.statusCode < 400) {
+      final location = rs.headers[HttpHeaders.locationHeader];
+
+      // sanity check for the header being present
+      if (location == null) {
+        logger.shout('Redirect response without location header.',
+            Exception('Redirect response without location header.'));
+        return rs;
+      }
```
Comment (jonasfj) [inline app/lib/shared/handler_helpers.dart:93]: Throw a 500 here, there is no reason to 303 when it clearly won't work.

File: app/lib/shared/handler_helpers.dart [hunk 2]
```diff
@@ -76,6 +77,50 @@ shelf.Handler wrapHandler(
...
+
+      // sanity check for the header being valid
+      final uri = Uri.tryParse(location);
+      if (uri == null) {
+        logger.shout(
+            'Redirect response with invalid location header: "$location".',
+            Exception(
+                'Redirect response with invalid location header: "$location".'));
+        return rs;
```
Comment (jonasfj) [inline app/lib/shared/handler_helpers.dart:102]: Same here, let's make this a 500

File: app/lib/shared/handler_helpers.dart [hunk 3]
```diff
@@ -76,6 +77,50 @@ shelf.Handler wrapHandler(
...
+      }
+
+      // path + querystring match
+      if (uri.toString().startsWith(uri.path) &&
+          request.requestedUri.toString().endsWith(uri.toString())) {
+        logger.shout(
+            'Redirect loop detected.', Exception('Redirect loop detected.'));
+        return rs;
+      }
```
Comment (jonasfj) [inline app/lib/shared/handler_helpers.dart:118]: This is if you do `pub.dev/path -> /path` ?

Comment (isoos) [reply app/lib/shared/handler_helpers.dart:113]: Yes, also matching it with `?queryString`.

**Review by jonasfj — APPROVED** [review]

File: app/lib/shared/handler_helpers.dart [hunk 4]
```diff
@@ -63,6 +63,7 @@ shelf.Handler wrapHandler(
   // Handlers wrap other handlers, and they are called in the revers order of
   // their wrapping. Read this list from the bottom and go up to get the real
   // execution order.
+  handler = _redirectLoopDetectorWrapper(logger, handler);
   handler = _cspHeaderWrapper(handler);
   handler = _userAuthWrapper(handler);
   handler =
```
Comment (jonasfj) [inline app/lib/shared/handler_helpers.dart:69]: @isoos 
unrelated nit: But should we perhaps use [shelf.Pipeline](https://pub.dev/documentation/shelf/latest/shelf/Pipeline-class.html) here?

Comment (isoos) [reply app/lib/shared/handler_helpers.dart:69]: Yeah, we are at the point where it would make sense :)

**Pull request merged**